// graphql-tools-fork/dist/stitching/proxiedResult.js
Object.defineProperty(exports, "__esModule", { value: true });
if ((typeof global !== 'undefined' && 'Symbol' in global) ||
    (typeof window !== 'undefined' && 'Symbol' in window)) {
    exports.SUBSCHEMAS_SYMBOL = Symbol('subschemas');
    exports.ERROR_SYMBOL = Symbol('subschemaErrors');
}
else {
    exports.SUBSCHEMAS_SYMBOL = Symbol('subschemas');
    exports.ERROR_SYMBOL = '@@__subschemaErrors';
}
function isProxiedResult(result) {
    return result && result[exports.ERROR_SYMBOL];
}
exports.isProxiedResult = isProxiedResult;
function getSubschemas(result) {
    return result && result[exports.SUBSCHEMAS_SYMBOL];
}
exports.getSubschemas = getSubschemas;
function setSubschemas(result, subschemas) {
    result[exports.SUBSCHEMAS_SYMBOL] = subschemas;
}
exports.setSubschemas = setSubschemas;
function setErrors(result, errors) {
    result[exports.ERROR_SYMBOL] = errors;
}
exports.setErrors = setErrors;
function getErrors(result, pathSegment) {
    var errors = result && result[exports.ERROR_SYMBOL];
    if (!Array.isArray(errors)) {
        return null;
    }
    var fieldErrors = [];
    for (var _i = 0, errors_1 = errors; _i < errors_1.length; _i++) {
        var error = errors_1[_i];
        if (!error.path || error.path[0] === pathSegment) {
            fieldErrors.push(error);
        }
    }
    return fieldErrors;
}
exports.getErrors = getErrors;
