// graphql-tools-fork/dist/transforms/ExtendSchema.js
/* tslint:disable:no-unused-expression */
Object.defineProperty(exports, "__esModule", { value: true });
var graphql_1 = require("graphql");
var generate_1 = require("../generate");
var MapFields_1 = require("./MapFields");
var ExtendSchema = /** @class */ (function () {
    function ExtendSchema(_a) {
        var typeDefs = _a.typeDefs, _b = _a.resolvers, resolvers = _b === void 0 ? {} : _b, defaultFieldResolver = _a.defaultFieldResolver, fieldNodeTransformerMap = _a.fieldNodeTransformerMap;
        this.typeDefs = typeDefs;
        this.resolvers = resolvers,
            this.defaultFieldResolver = defaultFieldResolver;
        this.transformer = new MapFields_1.default(fieldNodeTransformerMap);
    }
    ExtendSchema.prototype.transformSchema = function (schema) {
        this.transformer.transformSchema(schema);
        var newSchema;
        if (this.typeDefs) {
            newSchema = graphql_1.extendSchema(schema, graphql_1.parse(this.typeDefs));
        }
        return generate_1.addResolveFunctionsToSchema({
            schema: newSchema || schema,
            resolvers: this.resolvers,
            defaultFieldResolver: this.defaultFieldResolver,
        });
    };
    ExtendSchema.prototype.transformRequest = function (originalRequest) {
        return this.transformer.transformRequest(originalRequest);
    };
    return ExtendSchema;
}());
exports.default = ExtendSchema;
